File: output/GPT4o-Mini/xCodeEval_remove_4000/C++/Python/f849e1136527020734486d080f36ebc9.py
class IterRange:
    def __init__(self, first, second):
        self.first = first
        self.second = second

    def begin(self):
        return self.first

    def end(self):
        return self.second


def rev(origin):
    return IterRange(reversed(origin.end()), reversed(origin.begin()))


class IntIter:
    def __init__(self, i=0):
        self.i = i

    def __iter__(self):
        return self

    def __next__(self):
        result = self.i
        self.i += 1
        return result

    def __add__(self, step):
        return IntIter(self.i + step)

    def __sub__(self, step):
        return IntIter(self.i - step)

    def __eq__(self, other):
        return self.i == other.i

    def __ne__(self, other):
        return not self.__eq__(other)

    def __getitem__(self, n):
        return self.i + n


def range_func(b, e=None):
    if e is None:
        return IterRange(IntIter(0), IntIter(b))
    return IterRange(IntIter(b), IntIter(e if e > b else b))


def nrange(n):
    return range_func(1, n + 1)


def inrange(b, e):
    return range_func(b, e + 1)


def chmax(m, v):
    if v > m:
        m = v
        return True
    return False


def chmin(m, v):
    if v < m:
        m = v
        return True
    return False


def max_s(*args):
    return max(args)


def min_s(*args):
    return min(args)


class ListHead:
    def __init__(self):
        self.prev = self
        self.next = self

    def empty(self):
        return self.next == self

    def unlink(self):
        self.link(self.prev, self.next)
        self.reset()

    def insert_next(self, new_next):
        self.link(new_next, self.next)
        self.link(self, new_next)

    def insert_prev(self, new_prev):
        self.link(self.prev, new_prev)
        self.link(new_prev, self)

    @staticmethod
    def link(prev, next):
        prev.next = next
        next.prev = prev

    def reset(self):
        self.prev = self
        self.next = self


class ListIter:
    def __init__(self, head):
        self.cur = head

    def __iter__(self):
        return self

    def __next__(self):
        if self.cur is None:
            raise StopIteration
        result = self.cur
        self.cur = self.cur.next
        return result

    def __eq__(self, other):
        return self.cur == other.cur

    def __ne__(self, other):
        return not self.__eq__(other)


class Node:
    def __init__(self, v, pos):
        self.lh = ListHead()
        self.v = v
        self.pos = pos


maxn = 500005
lst = ListHead()
arr = [Node(0, i) for i in range(maxn)]
splay = []
ans = 0

n = int(input())
for i in range(n):
    v = int(input())
    arr[i].v = v
    arr[i].pos = i

for i in range(n):
    splay.append(arr[i])
    lst.insert_prev(arr[i].lh)

last_left = 0
last_right = 0

while splay:
    it = splay.pop(0)
    root = ListIter(lst)
    left = next(root).pos
    right = prev(root).pos
    if it.pos == left and it.pos == right:
        ans += min(last_left, last_right)
    elif it.pos == left:
        ans += last_left
        last_left = it.v
    elif it.pos == right:
        ans += last_right
        last_right = it.v
    else:
        ans += min(next(it).v, prev(it).v)
    it.lh.unlink()

print(ans)

File: output/GPT4o-Mini/xCodeEval_remove_4000/C++/Java/f849e1136527020734486d080f36ebc9.java
import java.util.*;

class IterRange<It> {
    private It first, second;

    public IterRange(It first, It second) {
        this.first = first;
        this.second = second;
    }

    public It begin() {
        return first;
    }

    public It end() {
        return second;
    }
}

class IntIter {
    private int i;

    public IntIter(int i) {
        this.i = i;
    }

    public int get() {
        return i;
    }

    public IntIter increment() {
        i++;
        return this;
    }

    public IntIter decrement() {
        i--;
        return this;
    }

    public IntIter add(int step) {
        return new IntIter(i + step);
    }

    public IntIter subtract(int step) {
        return new IntIter(i - step);
    }

    public int difference(IntIter that) {
        return i - that.i;
    }

    public boolean equals(IntIter that) {
        return i == that.i;
    }

    public boolean notEquals(IntIter that) {
        return !equals(that);
    }

    public int index(int n) {
        return i + n;
    }
}

class IntRange extends IterRange<IntIter> {
    public IntRange(IntIter first, IntIter second) {
        super(first, second);
    }
}

class Main {
    static long ans = 0;

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        Node[] arr = new Node[n];
        List<Node> lst = new LinkedList<>();
        SplayTree splay = new SplayTree();

        for (int i = 0; i < n; i++) {
            int v = scanner.nextInt();
            arr[i] = new Node(v, i);
        }

        for (Node node : arr) {
            splay.insert(node);
            lst.add(node);
        }

        int lastLeft = 0;
        int lastRight = 0;

        while (!splay.isEmpty()) {
            Node it = splay.begin();
            int left = lst.get(lst.indexOf(it) + 1).pos;
            int right = lst.get(lst.indexOf(it) - 1).pos;

            splay.erase(it);
            if (it.pos == left && it.pos == right) {
                ans += Math.min(lastLeft, lastRight);
            } else if (it.pos == left) {
                ans += lastLeft;
                lastLeft = it.v;
            } else if (it.pos == right) {
                ans += lastRight;
                lastRight = it.v;
            } else {
                ans += Math.min(lst.get(lst.indexOf(it) + 1).v, lst.get(lst.indexOf(it) - 1).v);
            }
            lst.remove(it);
        }

        System.out.println(ans);
    }
}

class Node {
    int v;
    int pos;

    public Node(int v, int pos) {
        this.v = v;
        this.pos = pos;
    }
}

class SplayTree {
    private Node root;

    public void insert(Node node) {
        // Insert logic for splay tree
    }

    public void erase(Node node) {
        // Erase logic for splay tree
    }

    public Node begin() {
        // Return the first node
        return root; // Placeholder
    }

    public boolean isEmpty() {
        return root == null;
    }

    public int size() {
        // Return size of the tree
        return 0; // Placeholder
    }
}
